src/vos/tests/evt_ctl.c

```diff
@@ -1440,6 +1448,169 @@ test_evt_iter_delete_internal(void **state)
+
+	for (count = 0; count < sizeof(val)/sizeof(int); count++) {
+		rc = evt_create(ts_feats, 13, arg->ta_uma, arg->ta_root,
+			DAOS_HDL_INVAL, &toh);
+		assert_int_equal(rc, 0);
+		rc = utest_sync_mem_status(arg->ta_utx);
+		assert_int_equal(rc, 0);
+		data_size = val[count];
+		D_PRINT("Data Size: %ld\n", data_size);
+		D_ALLOC(data, data_size);
+		strcpy(data, "EVTree: Out of Memory");
+		epr.epr_lo = 0;
+		/* Loop does the following : evt_insert,
+		* evt_find (first epoch) and evt_delete (random deletes)
+		* till out of space condition
+		*/
+		for (epoch = 1; ; epoch++) {
+			entry.ei_rect.rc_ex.ex_lo = epoch;
+			entry.ei_rect.rc_ex.ex_hi = epoch + data_size;
+			entry.ei_rect.rc_epc = epoch;
+			memset(&entry.ei_csum, 0, sizeof(entry.ei_csum));
+			entry.ei_ver = 0;
+			entry.ei_inob = data_size;
+			rc = bio_alloc_init(arg->ta_utx, &entry.ei_addr,
+					    data, data_size);
+			if (rc != 0) {
+				assert_int_equal(rc, -DER_NOSPACE);
+				break;
+			}
+			rc = evt_insert(toh, &entry);
+			if (rc != 0) {
+				assert_int_equal(rc, -DER_NOSPACE);
+				break;
+			}
+			rc = utest_check_mem_increase(arg->ta_utx);
+			assert_int_equal(rc, 0);
+			rc = utest_sync_mem_status(arg->ta_utx);
+			assert_int_equal(rc, 0);
+			if (epoch == 1) {
+				evt_ent_array_init(&ent_array);
+				extent.ex_lo = epoch;
+				extent.ex_hi = epoch + data_size;
+				epr.epr_hi = epoch;
+				rc = evt_find(toh, &epr, &extent, &ent_array);
+				if (rc != 0)
+					D_FATAL("Add rect failed %d\n", rc);
+				evt_ent_array_for_each(ent, &ent_array) {
+					static char *actual;
+
+					D_ALLOC(actual, data_size);
+					addr = ent->en_addr;
+					strncpy(actual,
+					(char *)utest_off2ptr(arg->ta_utx,
+						addr.ba_off),
+					(int)evt_extent_width(
+						&ent->en_sel_ext));
+					rc = strcmp(actual, data);
+					if (rc != 0)
+						fail_msg("Data Check Failed\n");
+					D_FREE(actual);
+				}
+				evt_ent_array_fini(&ent_array);
+			}
+			/* Delete a record*/
+			if (epoch % 10 == 0) {
+				entry.ei_rect.rc_ex.ex_lo = epoch;
+				entry.ei_rect.rc_ex.ex_hi = epoch + data_size;
+				entry.ei_rect.rc_epc = epoch;
+				rc = evt_delete(toh, &entry.ei_rect, NULL);
+				assert_int_equal(rc, 0);
+				rc = utest_check_mem_decrease(arg->ta_utx);
+				assert_int_equal(rc, 0);
+				rc = utest_sync_mem_status(arg->ta_utx);
+				assert_int_equal(rc, 0);
+			}
+		}
+		/* Delete remaining records: evt_iter_delete */
+		rc = evt_iter_prepare(toh, 0, NULL, &ih);
+		assert_int_equal(rc, 0);
+		rc = evt_iter_probe(ih, EVT_ITER_FIRST, NULL, NULL);
+		assert_int_equal(rc, 0);
+		print_message("Deleting evtree contents\n");
+		while (!evt_iter_empty(ih)) {
+			rc = evt_iter_delete(ih, NULL);
+			assert_int_equal(rc, 0);
+			rc = utest_check_mem_decrease(arg->ta_utx);
+			assert_int_equal(rc, 0);
+			rc = utest_sync_mem_status(arg->ta_utx);
+			assert_int_equal(rc, 0);
+		}
+		rc = evt_iter_finish(ih);
+		assert_int_equal(rc, 0);
+		/* Free allocated memory */
+		D_FREE(data);
```

Comment:
Need D_FREE(actual)

Reply:
Will add go to and free both data before any asserts (or) actual.